test/test_id_model.cpp

```diff
@@ -37,4 +38,350 @@ TEST_F(IdModelTest, DetectSelfMapping) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  auto tv1 = makeSymbolicTensor(2);
+  fusion.addInput(tv1);
+  auto tv2 = add(tv0, tv1);
+  fusion.addOutput(tv2);
+
+  auto tv3 = makeSymbolicTensor(2);
+  fusion.addInput(tv3);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+
+  // Merge adn split by one. The split input and output will be mapped.
+  for (auto tv : {tv0, tv1, tv2}) {
+    tv->merge(0)->split(0, 1);
+  }
+
+  // Also test an isolated trivial expr. Note that tv3 and tv4 are not
+  // connected with tv0, tv1 and tv2.
+  tv4->merge(0)->split(0, 1);
+
+  IdModel id_model(&fusion);
+  ValGraph vg = id_model.idGraph(IdMappingMode::EXACT);
+
+  // Map the split-by-1 input and output
+  vg.mapVals(tv2->axis(0), tv2->axis(0)->definition()->input(0));
+  vg.mapVals(tv4->axis(0), tv4->axis(0)->definition()->input(0));
+
+  ValGraphStmtSort vg_stmt_sort(vg);
+
+  std::vector<Expr*> ref_order;
+  ref_order.push_back(getParentExpr(tv2->axis(0), 2));
+  ref_order.push_back(getParentExpr(tv4->axis(0), 2));
+  ref_order.push_back(getParentExpr(tv2->axis(0), 1));
+  ref_order.push_back(getParentExpr(tv4->axis(0), 1));
+
+  checkSortingResults(vg, vg_stmt_sort.exprs(), vg_stmt_sort.vals(), ref_order);
+}
+
+// Sorting test with the same fusion as Indexing19
+TEST_F(IdModelTest, ValGraphStmtSort4) {
+  auto fusion = createFusionWithMultipleResolutionPaths();
+  FusionGuard fg(fusion.get());
+  auto all_tvs = ir_utils::allTvs(fusion.get());
+
+  // Since this fusion is not supported by ComputeAtMap, the
+  // validation flag must be false
+  IdModel id_model(fusion.get(), false, false, false);
+  id_model.buildExactGraph();
+  const ValGraph& vg = id_model.idGraph(IdMappingMode::EXACT);
+
+  ValGraphStmtSort vg_stmt_sort(vg);
+
+  auto tv1 = getTensorByName(all_tvs, 1);
+  auto tv2 = getTensorByName(all_tvs, 2);
+  auto tv4 = getTensorByName(all_tvs, 4);
+  auto tv5 = getTensorByName(all_tvs, 5);
+  auto tv6 = getTensorByName(all_tvs, 6);
+  auto tv8 = getTensorByName(all_tvs, 8);
+  auto tv9 = getTensorByName(all_tvs, 9);
+  auto tv10 = getTensorByName(all_tvs, 10);
+
+  // Expected reference order:
+  //
+  // exprg{39}: Merge iS2 bS3
+  // exprg{57}: Merge iS11 bS12
+  // exprg{17}: Merge iS17 bS18
+  // exprg{51 63}: Merge iS15 iS16
+  // exprg{69 73}: Split iS1
+  // exprg{9 25 33 45}: Merge iS20 iS21
+  // exprg{41}: Split iS46
+  // exprg{59}: Split iS61
+  // exprg{19}: Merge iS29 iS19
+  // exprg{53 65}: Split iS56
+  // exprg{71 75}: Split iS71
+  // exprg{11}: Merge iS23 iS22
+  // exprg{27}: Merge iS35 bS10
+  // exprg{35 47}: Split iS41
+  // exprg{43}: Split iS47
+  // exprg{61}: Split iS62
+  // exprg{21}: Split iS30
+  // exprg{55 67}: Split iS57
+  // exprg{13}: Split iS24
+  // exprg{29}: Split iS36
+  // exprg{37 49}: Split iS42
+  // exprg{23}: Split iS31
+  // exprg{15}: Split iS25
+  // exprg{31}: Split iS37
+
+  std::vector<Expr*> ref_order;
```

Comment:
IIUC, `ref_order` is **a** correct total order. Other total orders may be correct as well but just not produced due to the way IdModel is implemented. This may not be a big deal for now, but I can imagine people who later change the implementation of IdModel may run into such different-from-ref-but-still-correct problems. 

Have you considered to check only the partial order? E.g. for each def-use, check the def is ordered before the use. This would be the minimal requirement.

Reply:
Right, in fact that was my initial validation, but it's also important to have a deterministic order, so I switched to the current version. That's how I debugged the non-deterministic issue of #1714. 

We could also do the partial order validation when a failure with the total order validation is failed. That way, the failure message could say something like "an unexpected but valid order is detected" rather than just saying "an unexpected order is detected". I'll consider that if these tests create false alarms frequently.